man/battool.8
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH BATTOOL 8 "Sep 17, 2007"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)

.\" --------------------------------------------------------------------------
.\" Macro definitions
.\" --------------------------------------------------------------------------
.
.\" default indentation is 7, don't change!
.nr IN 5
.\" define indentation for suboptions
.nr SS 3
.\" add new suboption
.de IPs
.IP "\\$1" \n(SS 
..
.\" begin of first level suboptions, end with .RE
.de RSs
.RS \n(IN+3
..
.\" begin of 2nd level suboptions
.de RSss
.PD 0
.RS \n(SS+3
..
.\" end of 2nd level suboptions
.de REss
.RE 
.PD 1
..
.

.\" --------------------------------------------------------------------------

.TH "battool" 8
.SH NAME
battool - B.A.T.M.A.N. toolchain for B.A.T.M.A.N. Advanced
.SH SYNOPSIS
.B battool
.RI \fImodule\fP\ [\fIoptions\fP]\ {\fIdestination\ or\ interface\fP}
.br
.SH DESCRIPTION
B.A.T.M.A.N. Advanced operates on layer 2 and thus all hosts participating in the virtual switch are completely transparent for all protocols above layer 2. Therefore the common diagnosis tools do not work as expected. To overcome these problems the \fBB.A.T.M.A.N. toolchain (battool for short)\fP was created. At the moment the battool contains \fBping\fP, \fBtraceroute\fP and \fBbatdump\fP which can be used as module.
.PP
The battool uses the unix socket provided by B.A.T.M.A.N. Advanced to inject custom icmp packets into the data flow. Thats why ping and traceroute work almost like their IP based counterparts. Batdump analyzes the Traffic and packet types used B.A.T.M.A.N. Advanced.
.PP
.SH OPTIONS
.TP 
.RI \fBbattool\ ping\fP\ [\fIoptions\fP]\ [\fImac\ address|host\ name\fP]
The following options are supported:
.RSs
.IPs "-c (count)"
stop after n pings (default is infinite)
.IPs "-i (interval)"
interval to wait between 2 pings (default is 1 second)
.IPs "-t (timeout interval)"
timeout interval in seconds (default is 1 second)
.RE


.TP
.RI \fBbattool\ traceroute\fP\ [\fImac\ address|host name\fP]
Traceroute sends 3 packets to each hop, awaits the answers and prints out the response times.

.TP
.RI \fBbattool\ batdump\fP\ [\fIoptions\fP]\ [\fIinterface\fP]
The following options are supported:
.RSs
.IPs "-a (all packets)"
switch tcpdump behaviour to work as the "normal" tcpdump and filter for non-batman advanced packets
.IPs "-d (dump packets)"
print packet data in hex
.IPs "-p (packet type)"
specify to filter for certain batman packets:

.RSss 
.IPs "1" 
batman packets - these transport the B.A.T.M.A.N. originator messages to build the mesh network
.IPs "2"
batman icmp packets - lightweight implementation of the ICMP protocol (needed for ping and traceroute)
.IPs "3"
batman unicast packets - unicast data is bound as payload in these packets
.IPs "4"
batman broadcast packets - broadcast data is bound as payload in these packets
.RE

.IPs "-v (verbose)"
prints more information about each packet
.RE
.br
.SH FILES
.IPs bat-hosts
This file is simliar to the /etc/hosts file. You can write one MAC address and one host name per line. The battool will analyze the file to find the matching MAC address to your provided host name. Host names are much easier to remember than MAC addresses.  ;)
.SH SEE ALSO
\fIbatmand-adv\fP (8)
.SH AUTHOR
battool was written by Andreas Langer <a.langer@q-dsl.de>.
.PP
This manual page was written by Simon Wunderlich <siwu@hrz.tu-chemnitz.de>.


battool.c
/* Copyright (C) 2007 B.A.T.M.A.N. contributors:
 * Andreas Langer <a.langer@q-dsl.de>
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General Public
 * License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA
 *
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <errno.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>

#include "battool.h"

void usage() {
	printf("Usage:\n\tbattool -v\n\tbattool module [options] destination\n");
	printf("module: ping|batping|bp traceroute|batroute|br batdump|bd\n");
	printf("Use \"battool module -h\" for available options\n");
	exit(EXIT_FAILURE);
}

int compare_mac(void *data1, void *data2)
{
	return ( memcmp( data1, data2, sizeof(struct ether_addr) ) );
}

int choose_mac(void *data, int32_t size)
{
	unsigned char *key= data;
	uint32_t hash = 0, m_size = sizeof(struct ether_addr);
	size_t i;

	for (i = 0; i < m_size; i++) {
		hash += key[i];
		hash += (hash << 10);
		hash ^= (hash >> 6);
	}

	hash += (hash << 3);
	hash ^= (hash >> 11);
	hash += (hash << 15);

	return (hash%size);

}

void parse_hosts_file( struct hashtable_t *hash, char path[] ) {

	FILE *fd;
	char name[50], mac[18];
	struct ether_addr *tmp_mac;
	struct hosts *tmp_hosts;
	struct hashtable_t *swaphash;

	name[0] = mac[0] = '\0';

	if( ( fd = fopen(path, "r") ) == NULL )
		return;

	while( fscanf(fd,"%[^ \t]%s\n", name, mac ) != EOF ) {

		if( ( tmp_mac = ether_aton( mac ) ) == NULL ) {
			DBG("the mac address was not correct");
			return;
		}

		if( ( tmp_hosts = malloc(sizeof(struct hosts) ) ) == NULL ) {
			DBG("not enough memory for malloc");
			return;
		}


		memcpy(&tmp_hosts->mac, tmp_mac, sizeof(struct ether_addr));
		strncpy( tmp_hosts->name, name, 49 );

		hash_add(hash, tmp_hosts);

		if (hash->elements * 4 > hash->size) {

			swaphash = hash_resize(hash, hash->size * 2);

			if (swaphash == NULL) {

				printf("Couldn't resize hash table\n");

			}

			hash = swaphash;

		}

	}

	return;

}

int main( int argc, char **argv ) {
	int uid, ret=0;

	struct hashtable_t *host_hash;
	
	if( argc < 2 ) {
		usage();
	}

	if( strcmp( argv[1], "-v" ) == 0 ) {
		printf("Battool %s\n", VERSION);
		exit(EXIT_SUCCESS);
	}


	uid = getuid();
	if(uid != 0) { printf("You must have UID 0 instead of %d.\n",uid); exit(EXIT_FAILURE); }

	host_hash = hash_new( 64, compare_mac, choose_mac );

	parse_hosts_file( host_hash,HOSTS_FILE );
	
	if (strcmp(argv[1], "ping") == 0 ||strcmp(argv[1], "batping") == 0 || strcmp(argv[1], "bp") == 0 ) {
		/* call ping main function */
		ret = batping_main( argc-1, argv+1, host_hash );

	} else if(strcmp(argv[1], "traceroute") == 0 || strcmp(argv[1], "batroute") == 0 || strcmp(argv[1], "br") == 0  ) {
		/* call trace main function */
		ret = batroute_main( argc-1, argv+1, host_hash );

	} else if( strcmp(argv[1], "batdump") == 0 || strcmp(argv[1], "bd") == 0  ) {
		/* call trace main function */
		ret = batdump_main( argc-1, argv+1, host_hash );

	} else {

		usage();

	}

	hash_destroy(host_hash);
	return(ret);
}
